lib/Transforms/LLZKPCLLoweringPass.cpp

```diff
@@ -0,0 +1,422 @@
+  LogicalResult validateStruct(StructDefOp structDef) {
+    for (auto field : structDef.getFieldDefs()) {
+      auto fieldType = field.getType();
+      if (!llvm::dyn_cast<FeltType>(fieldType)) {
+        return structDef.emitError() << "Field must be felt type. Found " << fieldType
+                                     << " for field: " << field.getName();
+      }
+    }
+    return success();
+  }
+
+  /// Emit assertions for an equality `lhs == rhs`, with fast paths when one side
+  /// is a boolean and the other side is a constant {0,1}.
+  ///
+  /// Cases handled:
+  ///   - bool == 1  → assert(bool)
+  ///   - 1 == bool  → assert(bool)
+  ///   - bool == 0  → assert(!bool)
+  ///   - 0 == bool  → assert(!bool)
+  ///   - otherwise  → assert(lhs == rhs)
+  ///
+  /// Returns success after emitting IR.
+  static LogicalResult
+  emitAssertEqOptimized(OpBuilder &b, Location loc, Value lhsVal, Value rhsVal) {
+    // --- Small helpers --------------------------------------------------------
+    auto isBool = [](mlir::Value v) { return llvm::isa<pcl::BoolType>(v.getType()); };
+
+    auto getConstAPInt = [](Value v) -> std::optional<llvm::APInt> {
+      if (auto c = llvm::dyn_cast_or_null<pcl::ConstOp>(v.getDefiningOp())) {
+        // Chain: ConstOp -> FeltAttr (or BoolAttr-as-int) -> IntegerAttr -> APInt
+        return c.getValue().getValue().getValue();
+      }
+      return std::nullopt;
+    };
+
+    auto isConstOne = [&](mlir::Value v) {
+      if (auto ap = getConstAPInt(v)) {
+        return ap->isOne();
+      }
+      return false;
+    };
+    auto isConstZero = [&](mlir::Value v) {
+      if (auto ap = getConstAPInt(v)) {
+        return ap->isZero();
+      }
+      return false;
+    };
+
+    auto emitEqAssert = [&](mlir::Value l, mlir::Value r) {
+      auto eq = b.create<pcl::CmpEqOp>(loc, l, r);
+      b.create<pcl::AssertOp>(loc, eq.getRes());
+    };
+
+    auto emitAssertTrue = [&](mlir::Value pred) { b.create<pcl::AssertOp>(loc, pred); };
+
+    auto emitAssertFalse = [&](mlir::Value pred) {
+      auto neg = b.create<pcl::NotOp>(loc, pred);
+      b.create<pcl::AssertOp>(loc, neg.getRes());
+    };
+
+    // Optimized handling of boolean patterns
+    if (isBool(lhsVal) && isConstOne(rhsVal)) {
+      // bool == 1 → assert(bool)
+      emitAssertTrue(lhsVal);
+      return mlir::success();
+    }
+    if (isBool(rhsVal) && isConstOne(lhsVal)) {
+      // 1 == bool → assert(bool)
+      emitAssertTrue(rhsVal);
+      return mlir::success();
+    }
+    if (isBool(lhsVal) && isConstZero(rhsVal)) {
+      // bool == 0 → assert(!bool)
+      emitAssertFalse(lhsVal);
+      return mlir::success();
+    }
+    if (isBool(rhsVal) && isConstZero(lhsVal)) {
+      // 0 == bool → assert(!bool)
+      emitAssertFalse(rhsVal);
+      return mlir::success();
+    }
+
+    // Fallback to assert(lhs == rhs)
+    emitEqAssert(lhsVal, rhsVal);
+    return mlir::success();
```

Comment:
Nit: Since we're `using mlir;`, we can drop the namespace qualifications on the `mlir::success()` calls above.